### PR TITLE
feat: Get Performance Data & Performance Data Types

### LINKED ISSUE #361: Add Get Performance Data endpoint
## Description
Implement Get Performance Data endpoint to return the information of the system state which is supported to read as like cpu, memory, network traffic, and battery.

## Endpoint Documentation
Endpoint documentation: http://appium.io/docs/en/commands/device/performance-data/get-performance-data/

### PR DESCRIPTION
**Pull request by akinsolb**
closes #361

## Change list

Please provide briefly described change list which are you going to propose. 
 
## Types of changes

What types of changes are you proposing/introducing to .NET client?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] New feature (non-breaking change which adds value to the project)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Documentation
- [x] Have you proposed a file change/ PR with appium to update documentation? 
## Integration tests
- [x] Have you provided integration tests to pass against the beta version of appium? (for Bugfix or New feature)

## Details
Adds support for getting performance data and performance data types on Android
Adds the following methods to AndroidDriver.cs

GetPerformanceDataTypes()
GetPerformanceData(packageName,dataType,dataReadTimeout)

Please provide more details about changes if it is necessary. If there are new features you can provide code samples which show the way they
work and possible use cases. Also you can create [gists](https://gist.github.com) with pasted C# code samples or put them here using markdown. 
About markdown please read [Mastering markdown](https://guides.github.com/features/mastering-markdown/) and [Writing on GitHub](https://help.github.com/categories/writing-on-github/)

### CODE REVIEW

**Comment by akinsolb** [comment]
I couldn't get 'cpuinfo' to work/return a good response. An error occurred each time, so I commented out the assertion

**Comment by mykola-mokhnach** [comment]
@akinsolb Please provide the actual server error, perhaps I could fix it

File: src/Appium.Net/Appium/Android/AndroidDriver.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Android/AndroidDriver.cs:219]: What the `Support` prefix means here?

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:11]: Please enumerate the supported data types explicitly. Perhaps, it would make sense to have a separate enum for these

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:12]: what is the default read timeout?

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:33]: does zero value mean the timeout is not limited?

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:36]: It might be useful to provide an example of the output

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:11]: it makes sense to mention these data types are returned by `GetSupportedPerformanceDataTypes()` call

File: src/Appium.Net/Appium/Android/AndroidDriver.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Android/AndroidDriver.cs:219]: Ah, this is supposed to stated supported, but even that term is confusing, in order for someone to know what is supported they'd have to call GetPerformanceDataTypes.. thinking of making the datatype parameter an enum instead of a string

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:33]: I can't even remember the specification for the timeout. When I do, i'll add it to the xml documentation

File: src/Appium.Net/Appium/Android/AndroidDriver.cs
Comment (mykola-mokhnach) [reply src/Appium.Net/Appium/Android/AndroidDriver.cs:219]: I would say `GetPerformanceDataTypes` would be enough

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:11]: actually, these data types are static: https://github.com/appium/appium-android-driver/blob/c1a8ee54b69e14940d4c2ecb766be7f75235aba2/lib/commands/performance.js#L12

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:33]: actually, I've jus checked the source code. The argument name is confusing. The value means how many times the perf data retrieval operation should be retried in case of failure: https://github.com/appium/appium-android-driver/blob/c1a8ee54b69e14940d4c2ecb766be7f75235aba2/lib/commands/performance.js#L69

The default value is 2 with 1000ms delays between each retry

Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:33]: If I set this to 0 the appium logs doesn't display a second attempt, it just returns null.
1. I'm going to change the parameter to dataReadAttempts and update the param notes to make this clearer. 
2. Set the default value of dataReadAttempts to 1 because setting it to 0 returns null from the server (1.18).
3. Prevent users from setting 0 as the default value

**Comment by akinsolb** [comment]
> @akinsolb Please provide the actual server error, perhaps I could fix it

https://gist.github.com/akinsolb/78d3343b39baf2f989f8e5a21c8c4a52 This is a run on an emulator (Android 10)

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs:36]: The summary is an example of an output

File: src/Appium.Net/Appium/Android/AndroidDriver.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Android/AndroidDriver.cs]: done

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: made the above changes

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: done

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: I would rather create an overloaded method without this parameter. As far as I can see it is optional, so you could just skip sending it to the server: https://github.com/appium/appium-base-driver/blob/eba4635dcd8fe81899c3349d7e474acfb66037e7/lib/protocol/routes.js#L380

File: src/Appium.Net/Appium/Android/AndroidDriver.cs
Comment (mykola-mokhnach) [inline src/Appium.Net/Appium/Android/AndroidDriver.cs:221]: 👍

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: I could either create an overload or perform a check on the driver method code, as it's optional, it already works as an overload to users

**Comment by mykola-mokhnach** [comment]
I've pushed https://github.com/appium/appium-android-driver/pull/659 to fix the issue https://github.com/appium/appium-dotnet-driver/pull/422#issuecomment-687620106

You could check it out locally and verify now or wait until it's approved and is merged to beta

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (mykola-mokhnach) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: is it? We are always sending the `dataReadTimeout` arg in https://github.com/appium/appium-dotnet-driver/pull/422/files#diff-db1821172278290200c0e11b4a1a7b1dR120

**Comment by akinsolb** [comment]
> I've pushed [appium/appium-android-driver#659](https://github.com/appium/appium-android-driver/pull/659) to fix the issue [#422 (comment)](https://github.com/appium/appium-dotnet-driver/pull/422#issuecomment-687620106)
> 
> You could check it out locally and verify now or wait until it's approved and is merged to beta

Thanks. I'll wait till it's merged into beta.

File: src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs
Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: changes made

Comment (akinsolb) [reply src/Appium.Net/Appium/Interfaces/IHasPerformanceData.cs]: An alternative is making the dataReadAttempts argument nullable on the driver like so:

` public IList<object> GetPerformanceData(string packageName, string performanceDataType,
            int? dataReadAttempts = null) `
and checking for the value in the android command executor like so:

`public static object[] GetPerformanceData(IExecuteMethod executeMethod, string packageName, string dataType, int? dataReadTimeout)`

      `{
            var requiredKeys = new [] {"packageName", "dataType"};
            var requiredKeyValues = new object[] {packageName, dataType};
            if (dataReadTimeout is null)
                return executeMethod.Execute(AppiumDriverCommand.GetPerformanceData,
                    PrepareArguments(requiredKeys, requiredKeyValues)).Value as object[];
            requiredKeys.SetValue("dataReadTimeout", 2);
            requiredKeyValues.SetValue(dataReadTimeout, 2);
            return executeMethod.Execute(AppiumDriverCommand.GetPerformanceData,
                PrepareArguments(requiredKeys, requiredKeyValues)).Value as object[];
        }`

**Review by mykola-mokhnach — APPROVED** [review]
the cpuinfo fix has been published

**Pull request merged**
